// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: ext_config.proto

package unit

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on ExtConfig with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ExtConfig) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ExtConfig with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ExtConfigMultiError, or nil
// if none found.
func (m *ExtConfig) ValidateAll() error {
	return m.validate(true)
}

func (m *ExtConfig) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if utf8.RuneCountInString(m.GetAsset()) < 3 {
		err := ExtConfigValidationError{
			field:  "Asset",
			reason: "value length must be at least 3 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if utf8.RuneCountInString(m.GetAmount()) < 1 {
		err := ExtConfigValidationError{
			field:  "Amount",
			reason: "value length must be at least 1 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if m.GetIssuer() == nil {
		err := ExtConfigValidationError{
			field:  "Issuer",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if all {
		switch v := interface{}(m.GetIssuer()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, ExtConfigValidationError{
					field:  "Issuer",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, ExtConfigValidationError{
					field:  "Issuer",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetIssuer()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ExtConfigValidationError{
				field:  "Issuer",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return ExtConfigMultiError(errors)
	}

	return nil
}

// ExtConfigMultiError is an error wrapping multiple validation errors returned
// by ExtConfig.ValidateAll() if the designated constraints aren't met.
type ExtConfigMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ExtConfigMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ExtConfigMultiError) AllErrors() []error { return m }

// ExtConfigValidationError is the validation error returned by
// ExtConfig.Validate if the designated constraints aren't met.
type ExtConfigValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExtConfigValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExtConfigValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExtConfigValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExtConfigValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExtConfigValidationError) ErrorName() string { return "ExtConfigValidationError" }

// Error satisfies the builtin error interface
func (e ExtConfigValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExtConfig.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExtConfigValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExtConfigValidationError{}
